enhancement: Shared and Guaranteed CPU Pools

bartwensley opened the pull request:
This enhancement describes an approach to split the single CPU pool currently used by kubernetes into two partitions (shared and guaranteed). The guaranteed partition will be used exclusively for containers with the guaranteed QoS that consume whole CPUs. This will allow special kernel configuration to be applied to this set of CPUs, to enable low latency applications (e.g. 5G RAN vDU), without affecting other applications running on the same server.

bartwensley commented:
/cc @mrunalp

bartwensley commented:
/cc @MarSik @yanirq @ffromani @jmencak @Tal-or @swatisehgal @haircommander @rphillips

Tal-or commented on enhancements/shared-and-guaranteed-cpu-pools.md:
```diff
@@ -0,0 +1,407 @@
+* This enhancement assumes the feature is activated as part of installing the cluster and cannot be
+activated later. However, the sizing of the pools of shared and guaranteed CPUs can be modified (a reboot
+will be required).
+* The use of the shared and guaranteed CPU pools will be activated at the cluster level - it cannot be
```
yet, each node can has a slightly different allocation of shared and reserved right?

bartwensley replied:
The allocation of shared/reserved/isolated is done in the PerformanceProfile, which is linked to a MachineConfigPool. So to have different allocations of shared/reserved/isolated CPUs you would need to have multiple MachineConfigPools - the nodes in each pool would have the same allocations.

bartwensley replied:
Added a note to clarify that nodes can have different shared/guaranteed CPU allocations.

Tal-or commented on enhancements/shared-and-guaranteed-cpu-pools.md:
```diff
@@ -0,0 +1,407 @@
+* Update kubelet configuration to specify that the isolated CPUSet is to be used for guaranteed QoS
+containers using whole CPUs.
+* Update OpenShift Kubernetes API Server configuration to enable a new admission hook. Note: although this would
+work in single-node deployments, we still need to work through how enablement will work in multi-node deployments.
```
So as for now it should not work in MNO cluster?

bartwensley replied:
I think we should figure this out before the enhancement is completed. I wanted to get some general feedback on the approach before sorting out all the details.

browsell replied:
There should be no reason this can't work on a MNO

bartwensley replied:
Clarified that we will also consider multi-node deployments.

Tal-or commented on enhancements/shared-and-guaranteed-cpu-pools.md:
```diff
@@ -0,0 +1,407 @@
+
+The Kubelet (mostly the [CPU manager](https://github.com/openshift/kubernetes/tree/master/pkg/kubelet/cm/cpumanager) and its static policy) would be updated to:
+* Read and store the new `guaranteedCPUs` configuration.
+* Remove the `guaranteedCPUs` from the `defaultCpuSet` so that shared containers are excluded from the `guaranteedCPUs` when their affinity is set.
```
@swatisehgal @ffromani I wonder if this change requires changes in the way of how RTE/NFD preform CPU accounting when running under OCP and this feature is enabled.

browsell replied:
@Tal-or , can you elaborate on how this would affect RTE/NFD ?

Tal-or replied:
RTE/NFD is not aware to the internal pool allocation and it might affect the accounting that TAS scheduler is doing.

Let's have an example (this is long bear with me):
Assuming I've a cluster with:
1. TAS running (RTE/NFD + scheduler) 
2.  shared-partition feature enabled
3.  topology-manager-policy=single-numa-node (in Kubelet under the nodes)

I also have `NodeA` with the following available/free resources layout (this is how Kubelet sees it):
```yaml
NUMA0:
  guaranteed-cpus: 1
NUMA1:
  guaranteed-cpus: 1
  shared-cpus: 1
```
  
A new guaranteed pod is coming requesting for 2 integral CPUs (guaranteed CPUs) and some memory.
After it goes through the mutation hook the pod spec looks as follow:
```yaml
spec:
  resources:
    requests:
      cpu: 2000m
      memory: 100Mi
      openshift.io/guaranteed-cpus: 2000
    limits:
      cpu: 2000m
      memory: 100Mi
      openshift.io/guaranteed-cpus: 2000
```

When the pod gets to the default scheduler the scheduler marks  `NodeA` as suitable for hosting the pod because it sees two available guaranteed-cpus (the default scheduler is not aware of NUMA locality).

All the nodes that marked as suitable by the default scheduler are examine later by TAS scheduler.
TAS scheduler marks `NodeA` as suitable for hosting the pod as well because it sees two available CPUs under NUMA1 (TAS scheduler fed by RTE/NFD information and RTE/NFD is not aware of garanteed/shared CPUs division)

Finally the pod admitted to `NodeA` and reject by Kubelet with `topologyAffinityError` (becuase there're no 2 guaranteed CPUs that affine to the same NUMA) which is exactly the kind of cases that TAS meant to solve.

bartwensley replied:
Thanks @Tal-or - that does sound like an issue. So what are the options for addressing it? Would this require changes to the RTE/NFD to add the new shared/guaranteed extended resources to the NodeResourceTopology? I assume there is another component that reports the actual usage of the resources per NUMA node? And then would the TAS scheduler itself have to be updated to take these new resources into account? Is there a reasonable way to make this kind of scheduling enhancement?

browsell replied:
So it seems we would need to publish the avail cpus per numa

Tal-or replied:
Changes in RTE are technically possible, but NFD is more complex.
NFD is the u/s version that intended to replace RTE. 
Since NFD is an u/s component, I'm not sure it's possible to incorporate the changes needed to support this design.

bartwensley replied:
I added a new "Scheduler Awareness" open question and a new "Enhance Topology Aware Scheduler" alternative to capture the issues you raised in your comments. I still need to investigate this new option.

Tal-or commented on enhancements/shared-and-guaranteed-cpu-pools.md:
```diff
@@ -0,0 +1,407 @@
+  openshift.io/shared-cpus: 200
+limits:
+  cpu: 400m
+  openshift.io/shared-cpus: 200
```
This is really confusing.
If the scheduler looks only at the `requests` isn't that enough to only mutate the `requests` field?

Tal-or replied:
I think it worth mentioning in the enhancement then.

bartwensley replied:
Will do

Tal-or commented on enhancements/shared-and-guaranteed-cpu-pools.md:
```diff
@@ -0,0 +1,407 @@
+
+### Non-Goals
+
+* This enhancement assumes the feature is activated as part of installing the cluster and cannot be
```
The feature should be activated only when `spec.cpu.shared` specified in PerformanceProfile isn't it?

Tal-or replied:
So what part should be activated during installation and what part by specifying `spec.cpu.shared` in the profile?
I mean if user want to edit its profile and add `spec.cpu.shared` is it possible?

bartwensley replied:
The whole feature will need to be activated at install time (both the admission hook and the changes to kubelet). So spec.cpu.shared needs to be specified at installation time. It should be possible to changed the number of shared/guaranteed CPUs after installation, but not remove them (or add them on an already running system). The issue is that if you disable the feature, any existing pods with the new shared/guaranteed extended resources could not be scheduled. Likewise, if you tried to turn on the feature, any existing pods would not have the new shared/guaranteed extended resources, so they would be scheduled on nodes without doing the accounting for those resources and we would end up oversubscribing the shared or guaranteed CPUs on those nodes.

Maybe there is some way to address that?

Tal-or replied:
For the scheduling issue I think we can use the approach I described here and in the following comments:
https://github.com/openshift/enhancements/pull/1421#discussion_r1228388054 

But for making the feature dynamic you should be able to handle existing pods that are already running in the cluster, or pods that are not going through the scheduler (pods with `nodeName` or static pods) as well.
 
 Since the activation of the feature requires reboot, all the pods would go through the admission phase handled by Kubelet. Kubelet has all the information about running pods, so it can calculates for which pool each pod needs to be assign and set the CPU affinity respectively (similar to what CPU manager is doing today when `reservedSystemCPUs` value changes).

Now, if you're worried about oversubscribing to certain pool by the existing resources, we can use automated tools like PerformaceProfileCreator or others to compute the size of the pools for the user.

bartwensley replied:
I added a note here to indicate that the limitation of activating the feature only at installation time could be lifted depending on what the final design looks like.

Tal-or commented on enhancements/shared-and-guaranteed-cpu-pools.md:
```diff
@@ -0,0 +1,407 @@
+**cluster creator** is a human user responsible for deploying a cluster.
+
+1. The cluster creator creates a Performance Profile for the NTO and specifies a shared CPU partition.
+2. The cluster creator creates a machine config manifest to write a configuration file for kubelet to specify
```
Why this cannot be done by NTO?

bartwensley replied:
Right - I will update to indicate the manifest will be created by NTO.

Tal-or commented on enhancements/shared-and-guaranteed-cpu-pools.md:
```diff
@@ -0,0 +1,407 @@
+1. The cluster creator creates a Performance Profile for the NTO and specifies a shared CPU partition.
+2. The cluster creator creates a machine config manifest to write a configuration file for kubelet to specify
+the shared and isolated CPUs (TODO: or this could be the config option).
+3. The cluster creator (TODO: does what???) to enable the new admission hook.
```
commenting in order to track when this line is changing

bartwensley replied:
We're still sorting out whether we'll need the new admission hook, but enabling that would be at the cluster level (not the node level). Possibly by applying a CR to update the apiserver configuration, but not sure yet.

Tal-or commented on enhancements/shared-and-guaranteed-cpu-pools.md:
```diff
@@ -0,0 +1,407 @@
+be handled by the managementcpusoverride admission hook.
+   * a pod with Burstable or BestEffort QoS. The admission hook modifies the pod,
+adding `openshift.io/shared-cpus` requests/limits for each container, matching the CPU requests.
+   * a pod with Guaranteed QoS. The admission hook modifies the pod,
```
Just to make it more clear, I would dedicate a separate bullet for guaranteed pod with integral CPUs and guaranteed pod with non integral CPUs.

bartwensley replied:
Will do

Tal-or commented on enhancements/shared-and-guaranteed-cpu-pools.md:
```diff
@@ -0,0 +1,407 @@
+
+One constraint of this approach is that the feature scope is cluster wide. When the feature gate is
+enabled, the new admission hook will mutate all incoming pods, adding requests/limits for the new
+extended resources. This means that all nodes in the cluster must have a Performance Profile that
```
This is a very strict requirements.
This means that https://github.com/openshift/enhancements/pull/1396 tightly coupled with those requirements as well.

Tal-or commented on enhancements/shared-and-guaranteed-cpu-pools.md:
```diff
@@ -0,0 +1,407 @@
+
+### Open Questions [optional]
+
+1. How will this be enabled at the openshift apiserver level? It looks like you can create an admission hook 
```
Are opeshift apiserver pods going to be mutated as well? 
What about pods that are running before the admission hook starts running (if there're any)?

Tal-or replied:
> I will update the enhancement to be clear that the Management Workload Partitioning feature must be enabled in order to use this new feature

Decoupling it with WP means that even if the third pool would be accepted in u/s sometimes in the future, that still forces this feature to be OCP specific.

But if you are going in that direction I wonder if instead of having mutation hook, you can extend the default scheduler  using the scheduler-plugin framework  [https://github.com/kubernetes-sigs/scheduler-plugins](https://www.google.com/url?q=https://github.com/kubernetes-sigs/scheduler-plugins&sa=D&source=docs&ust=1686669446332969&usg=AOvVaw3WQj5R8Zguic5soxi1yOpI) 

The new scheduler plugin sees the guaranteed and shared cpus extended resources published by Kubelet, calculates the pod request based on the pod's QoS and CPUs requests/limts and admit the pod if the node has enough resources.

The extension for the scheduler is a plugin that runs as a pod, annotated with Management Workload Partitioning and placed on the reserved CPUs as well.

Tal-or replied:
The advantage of having an extension to the scheduler is that it's very common to do such thing u/s and it's completely pluggable and  dynamic.
IOW, you don't need that to be installed along with the cluster as you're doing with the admission hook.

bartwensley replied:
Thanks Tal-or - can you explain a bit more how this would work as a scheduler plugin? Are you suggesting that the pods would no longer have requests/limits for the extended resources? If that is the case, then how does the scheduler know how many shared/guaranteed CPUs are consumed already on each node by the pods currently running? Looking at the scheduler plugin interface it seems the interfaces just get passed the pod being scheduled, but I really don't know anything about this. Which plugin interfaces would we use for this and how would they be able to make the right decisions about where to schedule the new pod?

Tal-or replied:
> Are you suggesting that the pods would no longer have requests/limits for the extended resources?
Yes.

> If that is the case, then how does the scheduler know how many shared/guaranteed CPUs are consumed already on each node by the pods currently running? 

For every new pod that needs to be schedule, the scheduler would list all the existing pods on the node, calculates which pod belongs to which pool (By checking the QoS and resources of each pod).

At this point, the scheduler has the information about how much resources were allocated from each pool, and it also knows how much `Allocatable` resources at each pool the node has (because Kubelet populates this information to the node via the extended resources).

In the last step the scheduler figures out to which pool the new pod is belong (like it did for the existing pods), checks whether there are enough available resources, by subtract the allocated resources from `Allocatable` and mark the node as suitable or not.

> Looking at the scheduler plugin interface it seems the interfaces just get passed the pod being scheduled

You can initialize a lister to list all the non-terminal pods on the nodes.

> Which plugin interfaces would we use for this

https://pkg.go.dev/k8s.io/kubernetes/pkg/scheduler/framework#FilterPlugin

bartwensley replied:
Thanks!!! I will look into this option as it seems promising. One issue though is if we don't add the requests/limits for the new extended resources to the pods, the "[oc describe node](https://github.com/kubernetes/kubectl/blob/dfcbd3349ea39a580084e6c5c70897df13551071/pkg/util/resource/resource.go#L34)" won't show the "Allocated resources" properly. I guess we could patch the kubectl code to calculate the values for the shared/guaranteed extended resources based on the QoS of each pod, but that feels extremely ugly. I think we do need to give the user a way to see how much shared/guaranteed CPU resources are in use on each node, but maybe we don't want to use extended resources if we go with the scheduler plugin option? Any thoughts on a different/better way to show this?

Tal-or replied:
> the "[oc describe node](https://github.com/kubernetes/kubectl/blob/dfcbd3349ea39a580084e6c5c70897df13551071/pkg/util/resource/resource.go#L34)" won't show the "Allocated resources" properly.

hmm, I think you'll still be able to see it with `oc get -o yaml node` because it's part of the `NodeStatus`

> but maybe we don't want to use extended resources if we go with the scheduler plugin option

I don't see a way to completely avoid that, because the scheduler would need that information explicitly from the node. it needs to know the initial division of the different pools. 

Unless, you're passing that information to the scheduler when it starts. you can provide the pool size as an argument  to the scheduler, and the pool size consistent unless you change the value in the performanceProfile which requires reboot to the node anyway.

bartwensley replied:
> > the "[oc describe node](https://github.com/kubernetes/kubectl/blob/dfcbd3349ea39a580084e6c5c70897df13551071/pkg/util/resource/resource.go#L34)" won't show the "Allocated resources" properly.
> 
> hmm, I think you'll still be able to see it with `oc get -o yaml node` because it's part of the `NodeStatus`
> 

It won't be in the NodeStatus - it only contains the allocatable and the capacity. [The code](https://github.com/kubernetes/kubectl/blob/dfcbd3349ea39a580084e6c5c70897df13551071/pkg/util/resource/resource.go#L34) that sums up the  pod requests/limits to display "Allocated resources" depends on the extended resources requests/limits in each pod, which will not be there with this approach.

> > but maybe we don't want to use extended resources if we go with the scheduler plugin option
> 
> I don't see a way to completely avoid that, because the scheduler would need that information explicitly from the node. it needs to know the initial division of the different pools.
> 
> Unless, you're passing that information to the scheduler when it starts. you can provide the pool size as an argument to the scheduler, and the pool size consistent unless you change the value in the performanceProfile which requires reboot to the node anyway. That could work.

This won't work, because we can have multiple performanceProfiles (for different machine config pools).

Tal-or replied:
>  to display "Allocated resources" 

Oh sorry my bad I read it wrong.
Then yes you're right this is a bit more tricky. 
The only thing I can think of is using device-plugin to represent the shared/guaranteed as devices, but this is cumbersome.
Another option is to use new/existing object to contain this information.



> This won't work, because we can have multiple performanceProfiles (for different machine config pools).

No biggie. You can provide a ConfigMap that holds shared/guaranteed pool information associate with each node selector (the node selector map PerformanceProfile to group of  nodes).

The scheduler will examine each node based on the relevant pool (node -> node-selector -> shared/guaranteed pool)

bartwensley replied:
Hmm... I like the scheduler plugin approach because it feels much more self-contained. But the problem is going to be observability. Since we wouldn't be creating the new extended resources for shared/guaranteed CPUs, we have two issues:
1. How to publish the number of shared/guaranteed cpus configured on each node.
2. How to show the user how many shared/guaranteed cpus are in use on each node.

For 1, a simple option would be to have kubelet add annotations to each node (instead of publishing extended resources). The scheduler plugin could then use those annotations to know what was available on each node. Something like:
openshift.io/shared-cpus: x
openshift.io/guaranteed-cpus: y

For 2, we need a way for the user to see how many shared/guaranteed CPUs are configured and in use on each node. If extended resources are used (as in our current proposal), this is all automatic (handled already by "oc describe node"). If not using extended resources, I can think of a couple options:
1. We could add a new "oc" command to display the shared/guaranteed CPUs configured (by looking at the annotations on each node) and the shared/guaranteed CPUs in use (by looking at the pods on each node and their QoS class). 
2. We could patch the existing "oc describe node" command to add in this information (calculated in the same way). But this feels ugly and it doesn't look like we actually patch the kubectl code today (from what I can tell - not 100% sure though).

It doesn't feel like either of these options for showing the shared/guaranteed CPUs is going to be acceptable, so maybe going back to the approach of using extended resources and mutating the pods in the admission hook is the best we can do?

Tal-or replied:
> 2. How to show the user how many shared/guaranteed cpus are in use on each node.
We may kill two birds with one stone, if we'll manage to address https://github.com/openshift/enhancements/pull/1421#discussion_r1230103699 
Because RTE already presents in-used resources.

bartwensley replied:
I added a "New Scheduler Plugin" alternative to sum up this discussion.

Tal-or reviewed (commented):
There're some open questions in the enhancement, like how the admission hook is going to be enabled. 
I'll have  another review when those open questions would be addressed.

What OCP release is this feature planed for?

bartwensley commented:
> There're some open questions in the enhancement, like how the admission hook is going to be enabled. I'll have another review when those open questions would be addressed.
> 
> What OCP release is this feature planed for?

Thanks for the review Tal-or. We don't have a target release yet - need to make sure the proposed approach is acceptable first.

haircommander commented on enhancements/shared-and-guaranteed-cpu-pools.md:
```diff
@@ -0,0 +1,407 @@
+
+Here is the workflow:
+1. The cluster creator creates a Performance Profile for the NTO and specifies a shared CPU partition.
+2. The cluster creator creates a machine config manifest to configure CRI-O to partition shared workloads.
```
isn't this done by NTO as well?

bartwensley replied:
Yes - will update.

haircommander commented on enhancements/shared-and-guaranteed-cpu-pools.md:
```diff
@@ -0,0 +1,407 @@
+   * The scheduler sees the new pod and finds available `shared.workload.openshift.io/cores` resources on the node. 
+   The scheduler places the pod on the node.
+   * Kubelet does not do the cgroup configuration for the containers since it is a shared pod (based on the annotation).
+   * CRI-O does the cgroup configuration for the containers in the pod based on the shares/limits annotated 
```
maybe a note here saying this is the same workflow as solely guaranteed pods were in the past

bartwensley replied:
Yes - I will add that.

haircommander commented:
am I correct in seeing that the CRI-O behavior remains unchanged? the CPU shares and set are passed as annotation or by workload (respectively) by the kubelet and the change comes from the kubelet, scheduler and api handling the new kind of cpu?

bartwensley commented:
> am I correct in seeing that the CRI-O behavior remains unchanged? the CPU shares and set are passed as annotation or by workload (respectively) by the kubelet and the change comes from the kubelet, scheduler and api handling the new kind of cpu?

Correct - the proposed changes don't affect the existing CRI-O behavior. Thanks for your review.

bartwensley commented:
I added a new commit to address most of the comments that were added. I still need to investigate the TAS/RTE/NFD option further (just added it as a placeholder in the Alternatives section).

haircommander commented on enhancements/shared-and-guaranteed-cpu-pools.md:
```diff
@@ -0,0 +1,486 @@
+**cluster creator** is a human user responsible for deploying a cluster.
+
+1. The cluster creator creates a Performance Profile for the NTO and specifies a shared CPU partition.
+2. The cluster creator (TODO: does what???) to enable the new admission hook.
```
I can't remember if I've mentioned this already (I think I did but I can't see the history indicating I did) but I think NTO should do this and the cluster creator shouldn't do anything manual for it

bartwensley replied:
Yep - you raised this before. Here's my answer: https://github.com/openshift/enhancements/pull/1421#discussion_r1237300341

bartwensley replied:
Yes - I will do that with my next commit.

bartwensley replied:
Added a note as requested.

haircommander commented on enhancements/shared-and-guaranteed-cpu-pools.md:
```diff
@@ -0,0 +1,486 @@
+on the node based on the shared/isolated CPUSets specified in the config file.
+7. Something schedules:
+   * a pod with the `target.workload.openshift.io/management` annotation in a namespace 
+with the `workload.openshift.io/allowed` management annotation. The admission hook ignores the pod as it will
```
`The admission hook ignores the pod` which?

haircommander commented on enhancements/shared-and-guaranteed-cpu-pools.md:
```diff
@@ -0,0 +1,486 @@
+   * a pod with Burstable or BestEffort QoS. The admission hook modifies the pod,
+adding `openshift.io/shared-cpus` requests/limits for each container, matching the CPU requests.
+   * a pod with Guaranteed QoS. The admission hook modifies the pod as follows:
+     * for any containers with whole CPU requests/limits it adds `openshift.io/guaranteed-cpus` requests/limits
+     * for any containers with fractional CPU requests/limits it adds `openshift.io/shared-cpus` requests/limits
```
cc @Tal-or

bartwensley replied:
Thanks Peter - we'll need to work together to understand how the use of the "openshift.io/shared-cpus" by both features can be reconciled. I think there are some good options but this needs more investigation.

Tal-or replied:
Yes we can work it through once we get to the implementation.

haircommander commented:
what can we do to combine this and https://github.com/openshift/enhancements/pull/1396? They have similar aims and will ultimately share code, but the implementation is currently disjoint. Could we implement these as a progression where first shared are implemented, then mixed? or combine the two enhancements and implement both?

bartwensley commented on enhancements/shared-and-guaranteed-cpu-pools.md:
```diff
@@ -0,0 +1,486 @@
+that is disabled by default and then enabled through a config option on the API server - could that mechanism
+be used?
+
+#### Scheduler Awareness
+
+In the current proposal, the scheduler is aware of the new `openshift.io/shared-cpus` and
+`openshift.io/guaranteed-cpus` resources and will ensure that a pod is only scheduled on a node with enough
+of those resources available. However, this does not account for cases where the kubelet has a more
+restrictive Topology Manager Policy (e.g. `single-numa-node` policy). In that case, it is possible that a pod
+could be scheduled on a node that had enough total guaranteed CPUs (for example), but not enough guaranteed
+CPUs on a single NUMA node. This would result in kubelet rejecting the pod with a Topology Affinity error.
+
+One way to address this would be through the Topology Aware Scheduler and RTE/NFD - adding scheduler
+awareness of how many shared/guaranteed CPUs are available on each NUMA node. This will be explored further
+in the Topology Aware Scheduler alternative below.
+
+Note: In practice, this would only be an issue in multi-node deployments where there are pods that could
+run on a selection of nodes. In a single node deployment, the user will configure the number of shared and 
+guaranteed CPUs to match the workloads that they are planning on running on the node. The user would be 
+aware of the NUMA restrictions imposed by using the `single-numa-node` policy and would ensure their 
+configuration matched.
```
It turns out that scheduler awareness isn’t an issue (or at least isn’t any worse than it is without this enhancement). Prior to this enhancement, using the single-numa-node policy in kubelet’s Topology Manager can result in a Topology Affinity error when, for example, a container requests two guaranteed CPUs but the only two CPUs available are on two different NUMA nodes. The scenario is the same with the enhancement - the Topology Affinity error can occur if there are only two openshift.io/guaranteed-cpus available but they are on two different NUMA nodes.

If the user wants to avoid this scenario, the solution will still be the Topology Aware Scheduler (TAS). This should work with a very small change to the kubelet - when reporting the cpu resources available per NUMA node through the PodResources API, instead of reporting all available CPUs, when this shared/guaranteed CPU feature is enabled, we would only report the available guaranteed CPUs. This will ensure that the TAS will only schedule pods to nodes where there are enough guaranteed CPUs available on the same NUMA node (when the single-numa-node policy is being used).

Finally, after looking at the TAS/NFD/RTE, it isn’t going to be possible to use it as a solution to enable the shared/guaranteed CPU partition proposed in this enhancement. The main reason being that TAS/NFD/RTE can’t be deployed on some SNO configurations (e.g. vRAN DU) due to resource constraints. We need a solution that can be used in those configurations. A secondary reason is that the shared CPUs are not NUMA node specific - containers using shared CPUs float across all the shared CPUs - they are not allocated to specific containers so this doesn’t map well to the NRT and the role of the TAS. Tracking the shared/guaranteed CPUs as extended resources is much more intuitive and easily visible to the user (i.e. visible in “oc describe node” vs. the NRT CR).

Tal-or replied:
> instead of reporting all available CPUs, when this shared/guaranteed CPU feature is enabled, we would only report the available guaranteed CPUs.

Didn't think it through but at glance that could work.
And still TAS would have to learn about the shared-cpus quantity, but that can be done by reading the information from Kubelet. 

> The main reason being that TAS/NFD/RTE can’t be deployed on some SNO configurations (e.g. vRAN DU) due to resource constraints.

That's too bad.

bartwensley replied:
Updated the enhancement based on the comments above.

Tal-or commented:
> what can we do to combine this and #1396? They have similar aims and will ultimately share code, but the implementation is currently disjoint. Could we implement these as a progression where first shared are implemented, then mixed? or combine the two enhancements and implement both?

@haircommander I covered that in #1396  proposal and I saw your comments their. if you have anymore questions feel free to ask.

mrunalp commented on enhancements/shared-and-guaranteed-cpu-pools.md:
```diff
@@ -0,0 +1,486 @@
+scheduling pods to a particular node. Kubernetes currently has a single 
+[CPU resource](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#meaning-of-cpu) 
+that can be specified for each container. A change is necessary to ensure the scheduler will not over-allocate 
+shared or guaranteed CPUs on a node, resulting in a pod that would be scheduled but fail to run.
```
Do we have any additional detail on how to intend to solve the scheduler visibility?

mrunalp replied:
nvm, I see it described below.

mrunalp commented on enhancements/shared-and-guaranteed-cpu-pools.md:
```diff
@@ -0,0 +1,486 @@
+
+### Risks and Mitigations
+
+Nothing yet...
```
If this isn't upstream, then the additional risk of breakages and extending release timelines to address.

bartwensley replied:
I added this as a risk.

mrunalp commented:
I definitely see the use case that we are trying to address with this KEP. I want us to reduce our risk by having to carry a lot of downstream code for this and want us to see if we can upstream pieces of this and discuss this upstream as well.

bartwensley commented:
> I definitely see the use case that we are trying to address with this KEP. I want us to reduce our risk by not having to carry a lot of downstream code for this and want us to see if we can upstream pieces of this and discuss this upstream as well.

Thanks for the review Mrunal. I agree it would be good if we could upstream pieces of this - I'm just struggling to figure out what parts would be acceptable upstream. The basic building block here would be to enhance kubelet to distinguish between guaranteed and shared CPUs. But I don't see how that could be done without also tracking the shared/guaranteed CPUs. Tracking them with extended resources (as per the current proposal) wouldn't be acceptable upstream. The alternative of splitting the cpu resource into something like "cpu" and "cpu-guaranteed" also seems unacceptable as it would have huge impacts to existing code. Is there another way to do this?

mrunalp commented:
Based on the discussions we had, we need to include: 
1. Cgroup v2 enablement details
2. Feature flag 
3. CI plans to ensure we don't regress

bartwensley commented:
Thanks @mrunalp - I added another commit to address the discussions we had:

- Added details about using a feature flag
- Explored option of new cpumanager policy vs. modifying the existing static policy
- Discussed CI plans to ensure we don't regress

I didn't discuss the cgroup v2 enablement details as this enhancement doesn't change any of the cgroup related functionality - we should discuss cgroup v2 enablement as a separate topic.

mrunalp commented on enhancements/shared-and-guaranteed-cpu-pools.md:
```diff
@@ -88,8 +88,10 @@ kernel tuning applied to each pool.
 spec should be required.
-* The configuration should be optional and activated with a feature gate.
-* Although the initial usecase for this enhancement is single-node deployments, it should also be usable
+* The feature should be implemented behind a feature gate to ensure no impact to existing functionality.
```
Let's move this to a seprate Feature Gate section and also propose a name for it.

mrunalp replied:
nvm, I see the section below.

mrunalp commented:
@bartwensley Thanks for the updates. Let us follow up on v2 questions. We need to explicitly state in the enhancement that it will work with v2 and if it doesn't what are steps needed to get there.

bartwensley commented:
> Let us follow up on v2 questions. We need to explicitly state in the enhancement that it will work with v2 and if it doesn't what are the steps needed to get there.

I added a section to the "Implementation Details/Notes/Constraints" section specifying that the feature will work with cgroups v1 or v2 with no additional effort required.

mrunalp commented on enhancements/shared-and-guaranteed-cpu-pools.md:
```diff
@@ -0,0 +1,604 @@
+**cluster creator** is a human user responsible for deploying a cluster.
+
+1. The cluster creator creates a Performance Profile for the NTO and specifies a shared CPU partition.
+2. The cluster creator enables the new admission hook (TODO: how???). Note: This needs to be done
```
Can we clarify how we propose to do this? Could it be similar to the hook we have for  mgmt core pinning?

browsell replied:
The proposal would be to extend the admission plugin used for workload partitioning, https://github.com/openshift/kubernetes/tree/master/openshift-kube-apiserver/admission/autoscaling/managementcpusoverride

bartwensley replied:
After looking at it a bit more I think it would be best to create a new admission plugin and enable/disable it through the kube-apiserver config - I have made this update to the text.

mrunalp commented on enhancements/shared-and-guaranteed-cpu-pools.md:
```diff
@@ -0,0 +1,604 @@
+
+This feature is cgroup version agnostic - it does not require any changes to cgroup configuration for
+pods or containers. The feature will work with either cgroup v1 or v2, without modification to the
+code introduced by this feature.
```
Are there any additional difficulties introduced due to how load balancing is disabled in v2 now? See #1396 comments for how the cgroup heirarchy will need to be adjusted for that.

browsell replied:
No. This feature is agnostic to the cgroup version.
For containers requesting guaranteed cpus instead of allocating cpus from the defaultCpuset it is allocating cpus from a subset of the defaultCpuset. The runtime will just be given a set of cpus to be assigned to the container and would not impact how load balancing is being disabled.

bartwensley replied:
I updated this with some of the details from Brent's comment.

mrunalp commented:
Could we add conflicts with upstream projects such as VPA? If VPA is implemented for cpu manager then it won't work with this feature enabled due to the requirement of needing reboot to apply the kernel settings.

bartwensley commented:
> Could we add conflicts with upstream projects such as VPA? If VPA is implemented for cpu manager then it won't work with this feature enabled due to the requirement of needing reboot to apply the kernel settings.

I added a section on VPA to the "Implementation Details/Notes/Constraints" section - I don't believe there will be any issues with VPA interactions.

mrunalp reviewed (approved):
/lgtm
